### ISSUE TITLE
Can successfully build and deploy a release but auto-version and upgrades not working

### ISSUE THREAD

**Issue by nauger** [boldpoker/edeliver#88]
Admittedly, I'm new to Elixir and Phoenix, but first-- thank you for the work on this great deployment tool.

Some background:

We have an umbrella application, with a one app that really is the main application (a Phoenix/Coywboy endpoint.)  Our mix.exs lists this umbrella app at version 0.0.1.  Because this is a web-app with frequent (hopefully, many times a day) deploys, we do not want to have to increment and version control these versions.

Here are our exrm/edeliver deps:
```
{:edeliver, "~> 1.2.6"},
{:exrm, "~> 1.0.3"}
```

We can successfully build, deploy and then restart our application with:
```
mix edeliver build release
mix edeliver deploy release to production
mix edeliver restart production
```
That works great.  Unfortunately, we can't get the upgrade process with (AUTO_VERSION env) to work.  Here's our .deliver/config:
```
#!/usr/bin/env bash

APP="redline_web_app"

AUTO_VERSION=git-revision+branch-unless-master

BUILD_HOST="build.internal.com"
BUILD_USER="deploy"
BUILD_AT="/opt/redline/builds"
RELEASE_DIR="$BUILD_AT/apps/$APP/rel/$APP"

PRODUCTION_HOSTS="build.internal.com"
PRODUCTION_USER="deploy"
DELIVER_TO="/opt/redline"

pre_erlang_get_and_update_deps() {
  local _local_secret_config_file="apps/$APP/config/prod.secret.exs"
  if [ "$TARGET_MIX_ENV" = "prod" ]; then
    status "Copying '$_local_secret_config_file' file to build host"
    scp "$_local_secret_config_file" "$BUILD_USER@$BUILD_HOST:$BUILD_AT/apps/$APP/config/prod.secret.exs"
  fi
}

pre_erlang_clean_compile() {
  status "Running npm install"
  __sync_remote "
    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd \"$BUILD_AT/apps/$APP\"
    npm install $SILENCE
  "

  status "Running brunch build"
  __sync_remote "
    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd \"$BUILD_AT/apps/$APP\"
    node_modules/brunch/bin/brunch build --production $SILENCE
  "

  status "Running phoenix.digest"
  __sync_remote "
    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd \"$BUILD_AT\"
    if [ \"$BUILD_CMD\" = \"mix\" ]; then
      APP=\"$APP\" MIX_ENV=\"$TARGET_MIX_ENV\" $MIX_CMD do phoenix.digest $SILENCE
    fi
  "
}
```

And with this upgrade command:

```
mix edeliver upgrade production
```

Here is the result:

```
EDELIVER REDLINE_WEB_APP WITH UPGRADE COMMAND

-----> Upgrading to revision 0c097a9 from branch master
-----> Detecting release versions on production hosts
-----> Deploying upgrades to 1 online hosts
-----> Checking whether installed version 0.0.1 is in release store
-----> Building the upgrade from version 0.0.1
-----> Authorizing hosts
-----> Validating * version 0.0.1 is in local release store
-----> Ensuring hosts are ready to accept git pushes
-----> Pushing new commits with git to: deploy@build.internal.com
-----> Resetting remote hosts to 0c097a95d751400e877d9c5de7d8a27a60c2fbca
-----> Cleaning generated files from last build
-----> Checking out 0c097a95d751400e877d9c5de7d8a27a60c2fbca
-----> Copying 'apps/redline_web_app/config/prod.secret.exs' file to build host
-----> Fetching / Updating dependencies
-----> Running npm install
-----> Running brunch build
-----> Running phoenix.digest
-----> Compiling sources
-----> Checking version of new release
-----> Uploading archive of release 0.0.1 from local release store
-----> Extracting archive redline_web_app_0.0.1.tar.gz
-----> Detecting exrm version
-----> Generating release
-----> Removing built release 0.0.1 from remote release directory
ls: cannot access /opt/redline/builds/rel/redline_web_app/releases/*/: No such file or directory
Failed to detect generated release version at:
deploy@build.internal.com:/opt/redline/builds/rel/redline_web_app/releases/

The build upgrade has the same version as the installed version.
That release cannot be deployed as an upgrade. Alter the release version
or consider using the --auto-version=revision|commit-count|date option
to automatically increment the version or append the git revision.
It is recommended to enable the git-auto-revision permanently
for automatic upgrades by setting AUTO_VERSION=revision in the config.
```

Thanks for listening-- what am I missing?

**Comment by bharendt** [comment]
Thanks for the detailed description. I can't actually see any error in your configuration. 
The only thing to notice would be, that you actually would not need to specify the `RELEASE_DIR` in the edeliver config, it should be auto detected by edeliver.

But the error message and the build output could lead to the following conclusion:

Either the autoversioning does not work, even if you did set `AUTO_VERSION` env in your config correctly.
I think that should only happen if for any reason an older edeliver version without autoversioning support is used on the build host, the `AUTO_VERSION` env would be somehow overridden on the build host, or because the git revision could not be detected for any reason.

You could maybe validate the edeliver version on the build host and check what `git rev-parse HEAD` would return there and whether `echo $AUTO_VERSION` would return something on the build host.

__But more likely__, the `mix release` command __fails silently__ for some reason. This sometimes happens, depending on the exrm version. It might help to run the `upgrade` command with the `--verbose` option or when it failed, run `MIX_ENV=prod AUTO_VERSION="git-revision+branch-unless-master" mix release --verbosity=verbose` on the build host manually and carefully read the output and see where it stops. And then also check whether the release was built actually.

If the release was not built and the exrm `release` task did not output something like this:
```bash
The release for redline_web_app-0.0.1+git-hash is ready!
You can boot a console running your release with `$ rel/redline_web_app/bin/redline_web_app console`
```
some deeper investigation would be necessary why the release / upgrade cannot be build.

**Comment by nauger** [comment]
I've removed the unnecessary `RELEASE_DIR` from our config, but I agree, I think the culprit is auto versioning not working correctly.  On the build server, we have OTP 18.3 and Elixir 1.2.4.

When I run the following command, directly on the build server:
```
MIX_ENV=prod AUTO_VERSION="git-revision+branch-unless-master" mix release --verbosity=verbose
```
Eventually (after building all of the other apps in the umbrella, each of which have their default version of 0.0.1) I get:

```
===> Provider successfully run: resolve_release
===> Running provider release
===> Including Erts from /usr/lib64/erlang
===> release successfully created!
===> Provider successfully run: release
===> Running provider overlay
===> Provider successfully run: overlay
===> Running provider tar
===> tarball /opt/redline/builds/apps/redline_web_app/rel/redline_web_app/redline_web_app-0.0.1.tar.gz successfully created!
===> Provider successfully run: tar
==> Generating nodetool...
==> Packaging release...
==> The release for redline_web_app-0.0.1 is ready!
==> You can boot a console running your release with `$ rel/redline_web_app/bin/redline_web_app console`
```

**Comment by bharendt** [comment]
You are right. The auto-versioning does not work with umbrella project. 

The reason is that edeliver updates the release version of the project in the [`Mix.ProjectStack`](https://github.com/boldpoker/edeliver/blob/20b75b7d20a99a52dc48c3056b66623f354f4d25/lib/mix/tasks/release_version.ex#L86) before the `compile` task is executed which uses that version when generating the `.app` file for the application. 

At the moment this does not work for umbrella projects because they are not loaded when the edeliver `release version` task is executed.

So I tried the following to enable auto-versioning also for umbrella projects:

1) Using [`@recursive`](https://github.com/elixir-lang/elixir/blob/92f87b6b17b9272fc79e2f8fb94557805e6c2063/lib/mix/lib/mix/task.ex#L28) attribute for the edeliver [`release version` task](https://github.com/boldpoker/edeliver/blob/20b75b7d20a99a52dc48c3056b66623f354f4d25/lib/mix/tasks/release_version.ex#L1).

This enables the edeliver `release version` task to set the correct version for each umbrella app. But that config is dropped and a new one is generated again for each umbrella project when the `mix compile` task is executed (later). This behaviour is different from compiling a non-umbrella project.

So unfortunately this does not help.

2) Using an [alias](http://elixir-lang.org/docs/stable/mix/Mix.html) for `compile` in the umbrella project, e.g.

```elixir
  # apps/master_app

  def project do
    [app: :master_app,
     version: "0.0.1",
     deps_path: "../../deps",
     lockfile: "../../mix.lock",
     aliases: [compile: &update_version/1, "compile"]
     elixir: "~> 1.0",
     deps: deps]
  end

  defp update_version(_) do
    # ...
  end

```

This also does not work because the `compile` task is not invoked separately for the umbrella projects but only once for the main project. So the additional function used from the `aliases` section is never called.

3) Updating the version manually using the functions from the edeliver `release version` task, like:

```elixir
  import Mix.Tasks.Release.Version, only: [parse_args: 1, modify_version: 2]

  def project do
    [app: :master_app,
     version: parse_args("git-revision") |> modify_version("0.0.1"),
     deps_path: "../../deps",
     lockfile: "../../mix.lock",
     elixir: "~> 1.0",
     deps: deps]
  end

```

This is also not possible because the mix files are loaded before any dependencies, so the `Mix.Tasks.Release.Version` module is not available at that time.

4) So the __onliest option which would work for now__ is to update the version yourself in all umbrella project like this:

```elixir
  def project do
    [app: :master_app,
     version: modify_version("0.0.1"),
     deps_path: "../../deps",
     lockfile: "../../mix.lock",
     elixir: "~> 1.0",
     deps: deps]
  end

  def modify_version(version) do
    "#{version}+#{get_revision}"
  end

  defp get_revision() do
    System.cmd("git", ["rev-parse", "--short", "HEAD"])
    |> elem(0)
    |> String.rstrip
  end
```

**Comment by nauger** [comment]
Thank you for detailed run-down of options.  I have have amended our "primary" application (the one we edeliver with `APP="redline_web_app"`) to pull its version information from git as you suggested in **(4)**

This works and I can build versioned releases that are tagged with commits, and `mix edeliver version production` correctly reports the version+hash, as well as the branch and other commit meta data.

Unfortunately, I still can't use the `upgrade` task as it fails in a similar way:

```
-----> Upgrading to revision 38d4320 from branch master
-----> Detecting release versions on production hosts
-----> Deploying upgrades to 1 online hosts
-----> Checking whether installed version 0.0.1+f401647 is in release store
-----> Building the upgrade from version 0.0.1+f401647
-----> Authorizing hosts
-----> Validating * version 0.0.1+f401647 is in local release store
-----> Ensuring hosts are ready to accept git pushes
-----> Pushing new commits with git to: deploy@build.internal.com
-----> Resetting remote hosts to 38d4320b5bd4bb71e48418479c49b1c3ec582717
-----> Cleaning generated files from last build
-----> Checking out 38d4320b5bd4bb71e48418479c49b1c3ec582717
-----> Copying 'apps/redline_web_app/config/prod.secret.exs' file to build host
-----> Fetching / Updating dependencies
-----> Running npm install
-----> Running brunch build
-----> Running phoenix.digest
-----> Compiling sources
-----> Checking version of new release
-----> Uploading archive of release 0.0.1+f401647 from local release store
-----> Extracting archive redline_web_app_0.0.1+f401647.tar.gz
-----> Detecting exrm version
-----> Generating release
-----> Removing built release 0.0.1+f401647 from remote release directory
ls: cannot access /opt/redline/builds/rel/redline_web_app/releases/*/: No such file or directory
Failed to detect generated release version at:
deploy@build.internal.com:/opt/redline/builds/rel/redline_web_app/releases/
```

The directory that `ls` is trying to grab a release from does not have any sub directories:

```
-rw-rw-r-- 1 deploy deploy 4588 May 23 15:18 RELEASES
-rw-rw-r-- 1 deploy deploy   17 May 23 15:18 start_erl.data
```

**Comment by bharendt** [comment]
Does it work if you set the `RELEASE_DIR` to `="$BUILD_AT/apps/$APP/rel/$APP"` again? 
It looks like the detected release dir `/opt/redline/builds/rel/redline_web_app/releases/*/` is wrong.I think the release should be found at something like `/opt/redline/builds/app/redline_web_app/rel/redline_web_app/releases/*/` in your umbrella project.

**Comment by nauger** [comment]
Success!  (almost)  I added the following as suggested:

```
RELEASE_DIR="$BUILD_AT/apps/$APP/rel/$APP"
```

Unfortunately, the `erlang-build-upgrade` script [modifies this variable to:](https://github.com/boldpoker/edeliver/blob/master/strategies/erlang-build-upgrade#L151)
```
RELEASE_DIR="${DELIVER_TO}/rel/${APP}"
```

If I hardcode just that path to:

```
RELEASE_DIR="$MY_UMBRELLA_ORIGINAL_RELEASE_DIR"
```

### It works!  And I successfully get:

```
-----> Copying release 0.0.1+1826eb2 to local release store
-----> Copying redline_web_app.tar.gz to release store
-----> Upgrading production hosts to version 0.0.1+1826eb2
-----> Authorizing hosts
-----> Uploading archive of release 0.0.1+1826eb2 from local release store
copying .deliver/releases/redline_web_app_0.0.1+1826eb2.upgrade.tar.gz to deploy@build.internal.com:/opt/redline/redline_web_app/releases/0.0.1+1826eb2/redline_web_app_0.0.1+1826eb2.tar.gz
-----> Upgrading release to 0.0.1+1826eb2
Using /opt/redline/redline_web_app/releases/0.0.1+f401647/redline_web_app.sh
Release 0.0.1+1826eb2 not found, attempting to unpack releases/0.0.1+1826eb2/redline_web_app.tar.gz
Unpacked successfully: "0.0.1+1826eb2"
Generating vm.args/sys.config for upgrade...
sys.config ready!
vm.args ready!
Release 0.0.1+1826eb2 is already unpacked, now installing.
==> Checking whether upgrade to version 0.0.1+1826eb2 is possible...
---> Validating no process runs old code before upgrading...
---> None of 84 modules have old code.
---> Checking whether ranch socket acceptors can be found...
---> Found 100 ranch acceptors.
---> Checking whether ranch socket connections can be found...
---> Found 0 ranch connections.
---> Detected that no websockets are connected to channels.
==> Checking whether upgrade to version 0.0.1+1826eb2 is possible...
---> Validating no process runs old code before upgrading...
---> None of 84 modules have old code.
---> Checking whether ranch socket acceptors can be found...
---> Found 100 ranch acceptors.
---> Checking whether ranch socket connections can be found...
---> Found 0 ranch connections.
---> Detected that no websockets are connected to channels.
==> Preparing upgrade to version 0.0.1+1826eb2...
---> Suspending ranch socket acceptors...
---> Stopping 100 ranch socket acceptors...
---> Suspended 100 ranch acceptors.
---> Suspending phoenix websocket channels...
---> No websocket connections for phoenix channels are running.
---> Waiting up to 500 ms until running requests finished...
---> No requests running.
==> Upgrading release to version 0.0.1+1826eb2...
---> Upgrade to version 0.0.1+1826eb2 succeeded.
==> Resuming node after upgrade to version 0.0.1+1826eb2...
---> Notify running requests that new code was loaded...
---> No requests to notify.
---> Resuming phoenix websocket channels...
---> No websocket connections for phoenix channels are running.
---> Resuming ranch socket acceptors...
---> Starting 100 ranch socket acceptors...
---> Resumed 100 ranch acceptors.
==> Finished upgrade to version 0.0.1+1826eb2...
Installed Release: 0.0.1+1826eb2
Made release permanent: "0.0.1+1826eb2"
Version 0.0.1+1826eb2 is installed on host deploy@build.internal.com

UPGRADE DONE!
```

**Comment by bharendt** [comment]
Nice that it finally works. So thanks for your help to make the `upgrade` task to work also when using umbrella projects. 

I [updated](https://github.com/boldpoker/edeliver/commit/5d5fdf5d45d86428b666c0ec022d17935602ca4c) the `RELEASE_DIR` assignment [you found](https://github.com/boldpoker/edeliver/blob/466cedc935ef452130b9f385c26c87078708a0ec/strategies/erlang-build-upgrade#L151) to use the configured release directory from the `.deliver/config` as default. It is included into the [`1.2.7`](https://hex.pm/packages/edeliver/1.2.7) release.

This is the one and only situation where the `RELEASE_DIR` can't be auto-detected by edeliver because when using exrm, the old release must be copied and extracted to the correct (release) directory already before building, and detecting the release dir automatically is possible only after the build process.

So as a conclusion, I was wrong and __for umbrella projects you__

 __always have to set__ the `RELEASE_DIR` to `="$BUILD_AT/apps/$APP/rel/$APP"`

 because for each umbrella project a release is generated and it is not possible for edeliver to auto-detect the correct directory, especially when copying the old release (before building) which is required to build the upgrade.

 In addition the __auto-versioning does not work for umbrella projects__ and it needs to be added manually __to each umbrella project__ as in the example above.

I will add that information to a wiki page.

But I think it would be nice to find a way how to make auto-versioning work out of the box for umbrella project too, so I will keep this issue open. Maybe we also find a way how to auto-detect umbrella projects and the main application for the release.

**Comment by nauger** [comment]
Thanks for 1.2.7-- I just upgraded (and exrm just had a new release too.)  Unfortunately, this most recent upgrade (to a particular branch) with:

```
mix edeliver upgrade production --verbose --branch=edeliver-with-upgrade
```
Failed with:
```
This is an upgrade, verifying appups exist for updated dependencies..
==> Generated .appup for subapp1 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp2 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp3 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp4 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp5 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp6 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp7 0.0.1 -> 0.0.1+0a3d0dc
==> Generated .appup for subapp8 0.0.1 -> 0.0.1+0a3d0dc
==> All dependencies have appups ready for release!
==> Generated .appup for redline_web_app 0.0.1+1826eb2 -> 0.0.1+0a3d0dc
===> Starting relx build process ...
===> Resolving OTP Applications from directories:
          /opt/elixir-1.2.4/bin/../lib/eex/ebin
          /opt/elixir-1.2.4/bin/../lib/elixir/ebin
          /opt/elixir-1.2.4/bin/../lib/ex_unit/ebin
          /opt/elixir-1.2.4/bin/../lib/iex/ebin
          /opt/elixir-1.2.4/bin/../lib/logger/ebin
          /opt/elixir-1.2.4/bin/../lib/mix/ebin
          /opt/redline/builds/_build/prod/lib
          /opt/redline/builds/deps
          /opt/redline/builds/apps/redline_web_app/lib
          /usr/lib64/erlang/lib
          /opt/redline/builds/apps/redline_web_app/rel
===> Resolving available OTP Releases from directories:
          /opt/elixir-1.2.4/bin/../lib/eex/ebin
          /opt/elixir-1.2.4/bin/../lib/elixir/ebin
          /opt/elixir-1.2.4/bin/../lib/ex_unit/ebin
          /opt/elixir-1.2.4/bin/../lib/iex/ebin
          /opt/elixir-1.2.4/bin/../lib/logger/ebin
          /opt/elixir-1.2.4/bin/../lib/mix/ebin
          /opt/redline/builds/_build/prod/lib
          /opt/redline/builds/deps
          /opt/redline/builds/apps/redline_web_app/rel
===> Resolved redline_web_app-0.0.1+0a3d0dc
===> Including Erts from /usr/lib64/erlang
===> release successfully created!
==> ERROR: "Failed to build release. Please fix any errors and try again."

A remote command failed on:

  deploy@build.internal.com

Output of the command is shown above and the command executed
on that host is printed below for debugging purposes:

FAILED with exit status 1:

    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd /opt/redline/builds
    if [ "mix" = "rebar" ]; then
      echo "using rebar to generate release"
      ./rebar   -f generate
    elif [ "mix" = "relx" ]; then
      echo "using relx to generate release"
      ./relx release
    elif [ "mix" = "mix" ]; then
      echo "using mix to generate release"
      MIX_ENV="prod" LINK_SYS_CONFIG="" LINK_VM_ARGS="" APP="redline_web_app" AUTO_VERSION="" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do release.version  --verbose,  release  --no-confirm-missing --verbosity=normal
    fi
```

**Comment by bharendt** [comment]
The error output:
```
==> ERROR: "Failed to build release. Please fix any errors and try again."
```
is from exrm, while the success message 
```
===> release successfully created!
```
printed in the line before is from relx. 

Maybe you could retry with the `exrm` version you used before and check whether it works with that version.

It looks like the new exrm version `1.0.5` will also [break](https://github.com/bitwalker/exrm/commit/6df2765b27e29dc910be959a49385951d68d83f9) the [node commands](https://github.com/boldpoker/edeliver/blob/ce77a0675a364b43c50d3b90e9a3ba4e85f94545/strategies/erlang-node-execute#L82) because the first line of output was removed.

**Comment by nauger** [comment]
You're right!  Upgrading `exrm` was misguided.  Rolling back to `1.0.4` fixed the issue and upgrades are building once more.  Additionally, I was scratching my head about why the node command output had disappeared, so thank you for that too!

**Comment by nauger** [comment]
You're right!  Upgrading `exrm` was misguided.  Rolling back to `1.0.4` fixed the issue and upgrades are building once more.  Additionally, I was scratching my head about why the node command output had disappeared, so thank you for that too!

**Comment by nauger** [comment]
This might be better suited as a new issue (or perhaps more related to exrm) but I'll post here for continuity.  So, given the above, we have an umbrella app, where each app sets it's version based on the git revision.  So, we expect all of the umbrella apps to have the exact same version when we upgrade.

```
redline_web_app: 0.0.1+xyz123
app1: 0.0.1+xyz123
app2: 0.0.1+xyz123
app3: 0.0.1+xyz123
...
```

In the the verbose output, you can see each of these apps being built, with the correct versioned name, but when the main app is finally released and deployed, the sub-apps still report their old version number.

````
mix edeliver version production
```

Shows the correct version for the "master" application, but using the remote console:

```
./redline_web_app remote_console
Using /opt/redline/redline_web_app/releases/0.0.1+669c9e6/redline_web_app.sh
Erlang/OTP 18 [erts-7.3] [source-d2a6d81] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Interactive Elixir (1.2.4) - press Ctrl+C to exit (type h() ENTER for help)
iex(redline_web_app@build)1> :application.get_key(:redline_web_app, :vsn)
{:ok, '0.0.1+669c9e6'}
iex(redline_web_app@build)2> :application.get_key(:app1, :vsn)
{:ok, '0.0.1+bb33ca1'}
```

The sub app is still running the previous version (bb33ca1).

**Comment by bharendt** [comment]
Does `:application.which_applications` and `:release_handler.which_releases` also report the old versions for the umbrella apps? If yes, it would be interesting whether they are loaded in the `relup` instructions of the upgrade.

**Comment by nauger** [comment]
Yes-- both of these report the new version of the `master` app, and the old version of all other apps in the umbrella.

**Comment by bharendt** [comment]
Are the new versions of the umbrella apps included at all into the upgrade tar at `redline_web_app _0.0.1+669c9e6.upgrade.tar.gz:lib/app1-0.0.1+669c9e6` and then in the running release at `redline_web_app/lib/app1-0.0.1+669c9e6` etc. respectively? And they need to be loaded in the `redline_web_app/releases/0.0.1+669c9e6/relup` file. If not, you can maybe try to build an upgrade using the `--skip-relup-mod` and check the `relup` file again.

**Comment by nauger** [comment]
The new versions of the sub-apps are not listed in the relup.  Here's the command I just ran:

```
mix edeliver upgrade production --verbose --skip-relup-mod
```

Which resulted in:
```
This is an upgrade, verifying appups exist for updated dependencies..
==> Generated .appup for app1 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app2 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app3 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app4 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app5 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app6 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app7 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app8 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> All dependencies have appups ready for release!
==> Generated .appup for redline_web_app 0.0.1+669c9e6 -> 0.0.1+7a703d0
===> Starting relx build process ...
===> Resolving OTP Applications from directories:
          /rz/elixir-1.2.4/bin/../lib/eex/ebin
          /rz/elixir-1.2.4/bin/../lib/elixir/ebin
          /rz/elixir-1.2.4/bin/../lib/ex_unit/ebin
          /rz/elixir-1.2.4/bin/../lib/iex/ebin
          /rz/elixir-1.2.4/bin/../lib/logger/ebin
          /rz/elixir-1.2.4/bin/../lib/mix/ebin
          /opt/redline/builds/_build/prod/lib
          /opt/redline/builds/deps
          /opt/redline/builds/apps/redline_web_app/lib
          /usr/lib64/erlang/lib
          /opt/redline/builds/apps/redline_web_app/rel
===> Resolving available OTP Releases from directories:
          /rz/elixir-1.2.4/bin/../lib/eex/ebin
          /rz/elixir-1.2.4/bin/../lib/elixir/ebin
          /rz/elixir-1.2.4/bin/../lib/ex_unit/ebin
          /rz/elixir-1.2.4/bin/../lib/iex/ebin
          /rz/elixir-1.2.4/bin/../lib/logger/ebin
          /rz/elixir-1.2.4/bin/../lib/mix/ebin
          /opt/redline/builds/_build/prod/lib
          /opt/redline/builds/deps
          /opt/redline/builds/apps/redline_web_app/rel
===> Resolved redline_web_app-0.0.1+7a703d0
===> Including Erts from /usr/lib64/erlang
===> release successfully created!
===> relup successfully created!
===> tarball /opt/redline/builds/apps/redline_web_app/rel/redline_web_app/redline_web_app-0.0.1+7a703d0.tar.gz successfully created!
==> The release for redline_web_app-0.0.1+7a703d0 is ready!
==> You can boot a console running your release with `$ rel/redline_web_app/bin/redline_web_app console`
-----> Removing built release 0.0.1+669c9e6 from remote release directory
-----> Copying release 0.0.1+7a703d0 to local release store
-----> Copying redline_web_app.tar.gz to release store
-----> Upgrading production hosts to version 0.0.1+7a703d0
-----> Authorizing hosts
-----> Uploading archive of release 0.0.1+7a703d0 from local release store
copying .deliver/releases/redline_web_app_0.0.1+7a703d0.upgrade.tar.gz to deploy@build.internal.com:/opt/redline/redline_web_app/releases/0.0.1+7a703d0/redline_web_app_0.0.1+7a703d0.tar.gz
-----> Upgrading release to 0.0.1+7a703d0
Using /opt/redline/redline_web_app/releases/0.0.1+669c9e6/redline_web_app.sh
Release 0.0.1+7a703d0 not found, attempting to unpack releases/0.0.1+7a703d0/redline_web_app.tar.gz
Unpacked successfully: "0.0.1+7a703d0"
Generating vm.args/sys.config for upgrade...
sys.config ready!
vm.args ready!
Release 0.0.1+7a703d0 is already unpacked, now installing.
Installed Release: 0.0.1+7a703d0
Made release permanent: "0.0.1+7a703d0"
Version 0.0.1+7a703d0 is installed on host deploy@build.internal.com

UPGRADE DONE!
```

The final `relup` only shows the changes to the master application, and not version changes to the subs.

**Comment by bharendt** [comment]
Not sure. Maybe the umbrella apps are not updated because they did not change. 
It would be necessary to check the relx / release_handler code which generates the `relup` file to verify that.

But maybe they are also omitted because they are upgraded from a different version. 
You output shows that app1 - app8 are upgraded from `0.0.1+bb33ca1` while your main 
app is upgraded from `0.0.1+669c9e6` to `0.0.1+7a703d0`.

```
...
==> Generated .appup for app7 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> Generated .appup for app8 0.0.1+bb33ca1 -> 0.0.1+7a703d0
==> All dependencies have appups ready for release!
==> Generated .appup for redline_web_app 0.0.1+669c9e6 -> 0.0.1+7a703d0
```

**Comment by bharendt** [comment]
Closing this because building releases for umbrella apps will be more easy and less error-prone when using [distillery](https://github.com/bitwalker/distillery) which will be the default for the next [major release](https://github.com/boldpoker/edeliver/milestone/1).
